File: go/rates.go
package walletpay

import (
	"context"

	"github.com/openweb3-io/wallet-pay-openapi/go/internal/openapi"
)

type (
	EstimateResponse = openapi.EstimateResponse
	GetRatesRequest  = openapi.GetRatesRequest
	GetRatesResponse = openapi.GetRatesResponse
	CurrencyPair     = openapi.CurrencyPair
)

type Rates struct {
	api *openapi.APIClient
}

type EstimateOptions struct {
	BaseCurrency string
	BaseAmount   string
	ToCurrency   string
}

func (e *Rates) Estimate(ctx context.Context, options *EstimateOptions) (*EstimateResponse, error) {
	req := e.api.RatesApi.V1RatesEstimate(ctx)
	req = req.BaseCurrency(options.BaseCurrency)
	req = req.BaseAmount(options.BaseAmount)
	req = req.ToCurrency(options.ToCurrency)
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := EstimateResponse(out)
	return &ret, nil
}

func (e *Rates) GetRates(ctx context.Context, getRatesIn *GetRatesRequest) (*GetRatesResponse, error) {
	req := e.api.RatesApi.V1RatesList(ctx)
	req = req.GetRatesRequest(openapi.GetRatesRequest(*getRatesIn))
	out, res, err := req.Execute()
	if err != nil {
		return nil, wrapError(err, res)
	}
	ret := GetRatesResponse(out)

	return &ret, nil
}

File: java/lib/src/main/java/io/openweb3/walletpay/Rates.java
package io.openweb3.walletpay;

import io.openweb3.walletpay.exceptions.ApiException;
import io.openweb3.walletpay.internal.api.RatesApi;
import io.openweb3.walletpay.models.EstimateResponse;
import io.openweb3.walletpay.models.GetRatesRequest;
import io.openweb3.walletpay.models.GetRatesResponse;

public final class Rates {
	private final RatesApi api;

	Rates() {
		api = new RatesApi();
	}

	public EstimateResponse estimate(final RatesOptions options) throws ApiException {
		try {
			return api.v1RatesEstimate(options.getBaseCurrency(), options.getBaseAmount(), options.getToCurrency());
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}

	public GetRatesResponse getRates(final GetRatesRequest ratesIn) throws ApiException {
		try {
			return api.v1RatesList(ratesIn);
		} catch (io.openweb3.walletpay.internal.ApiException e) {
			throw Utils.WrapInternalApiException(e);
		}
	}
}
